vcd/resource_vcd_dnat.go

```diff
@@ -220,31 +247,96 @@ func resourceVcdDNATDelete(d *schema.ResourceData, meta interface{}) error {
 	}
 
 	edgeGateway, err := vcdClient.GetEdgeGatewayFromResource(d, "edge_gateway")
+	if err != nil {
+		return fmt.Errorf(errorUnableToFindEdgeGateway, err)
+	}
+
+	if "" != d.Get("network_name").(string) {
+		err = edgeGateway.RemoveNATRule(d.Id())
+		if err != nil {
+			return fmt.Errorf("error deleting SNAT rule: %#v", err)
+		}
+	} else {
+		// this for back compatibility when network name and network type isn't provided - TODO remove with major release
+		task, err := edgeGateway.RemoveNATPortMapping("DNAT",
+			d.Get("external_ip").(string),
+			portString,
+			d.Get("internal_ip").(string),
+			translatedPortString)
+		if err != nil {
+			return fmt.Errorf("error setting DNAT rules: %#v", err)
+		}
+
+		err = task.WaitTaskCompletion()
+		if err != nil {
+			return fmt.Errorf("error completing tasks: %#v", err)
+		}
+	}
+	return nil
+}
+
+func resourceVcdDNATUpdate(d *schema.ResourceData, meta interface{}) error {
 
+	vcdClient := meta.(*VCDClient)
+
+	// Update supports only when network name and network type provided
+	networkName := d.Get("network_name")
+	if nil == networkName || networkName.(string) == "" {
+		return fmt.Errorf("update works only when network_name and network_type is provided and rule created using them \n")
+	}
+
+	edgeGateway, err := vcdClient.GetEdgeGatewayFromResource(d, "edge_gateway")
 	if err != nil {
 		return fmt.Errorf(errorUnableToFindEdgeGateway, err)
 	}
-	task, err := edgeGateway.RemoveNATPortMapping("DNAT",
-		d.Get("external_ip").(string),
-		portString,
-		d.Get("internal_ip").(string),
-		translatedPortString)
+
+	natRule, err := edgeGateway.FetchNatRule(d.Id())
 	if err != nil {
-		return fmt.Errorf("error setting DNAT rules: %#v", err)
+		log.Printf("Error: Nat rule isn't found")
+		d.SetId("")
+		return err
```

Comment:
Could this error give more context? I've checked `FetchNatRule` and it doesn't return the ID or more details as well.

Reply:
Sorry, didn't get question...

Reply:
I mean the `return err`. This error goes to user but it doesn't contain any context. Like nat rule ID or something more.

Reply:
Same applies to log.Printf("Error: Nat rule isn't found"). It doesn't give answer to "which nat rule isn't found"

Reply:
I explained in other Tread

Reply:
What @Didainius is suggesting is return more context, such as `return fmt.Errorf(" rule %s not found: %s", d.Id(), err)`

Reply:
I understood what Dainius suggested, my concern was other thing. Overall I pushed change, what logging and resouce is removed from state if not found.